Add support for building S.P.CoreLib in a separate job

mangod9 opened the pull request:
**WIP** to add support to build System.Private.CoreLib separately so coreclr and libraries builds can run in parallel.

mangod9 committed:
Add support for building S.P.CoreLib in a separate job

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=34166) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=34166) before we can accept your contribution.<br/><br/>:x: mangod9 [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=34166)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=34166) it.</sub>

mangod9 commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,163 @@
+    pool: ${{ parameters.pool }}
+    condition: ${{ parameters.condition }}
+
+    # Compute job name from template parameters
```
wonder if these options are required for corelib

jkoritzinsky replied:
We still need the names for osGroup, arch, and config for sure. We may still need osSubgroup. We don't need compiler though.

Also, can we change the job display names to say that these are CoreLib jobs? Makes it a lot easier to read on the AzDO view.

mangod9 replied:
cleaned up some parameters and also updated display names.

mangod9 commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,163 @@
+
+    steps:
+
+    # Install native dependencies
```
native dependencies might not be required.

jkoritzinsky replied:
We need to make sure python is installed, so we probably still need this.

mangod9 commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,163 @@
+    - script: $(Build.SourcesDirectory)$(dir)build$(scriptExt) -subsetCategory coreclr -subset corelib $(crossArg) -arch $(archType) -c $(buildConfig) $(officialBuildIdArg) -ci
+      displayName: Build System.Private.CoreLib only
+
+    # Get key vault secrets for publishing
```
is keyvault secret required to upload artifacts?

jkoritzinsky replied:
I think we can remove this step.

safern replied:
No those are old stuff from coreclr, we can remove the vault secrets stuff, they were used for publishing

mangod9 replied:
removed keyvault step.

mangod9 commented on eng/pipelines/runtime.yml:
```diff
@@ -118,6 +118,44 @@ jobs:
       - eng/pipelines/mono/*
       - eng/pipelines/libraries/*
 
+#
```
are both checked and retail flavors required for corelib, or are they essentially the same in both flavors?

jkoritzinsky replied:
The CoreLib flavor has to match the CoreCLR flavor.

safern replied:
Note that you can use `platform-matrix-multijob` to make this simpler. You can do something like:

```yml
- template: /eng/pipelines/common/platform-matrix-multijob.yml
  parameters:
    platforms:
    - Windows_NT_x64
    - Linux_x64
   jobTemplates:
   - jobTemplate: /eng/pipelines/coreclr/templates/build-corelib-job.yml
     buildConfig: checked
     jobParameters:
       testGroup: inerloop
   - ...
    More job templates
   - ...
  
```

That way you share the platform matrix across multiple jobs to keep in sync.

mangod9 replied:
thanks for the tip, will update once we get the basic parallelization working.

mangod9 replied:
I am going to defer this optimization to a subsequent change.

jkoritzinsky commented:
It looks like we're missing a step to upload the intermediate artifacts so they can get pulled down by later jobs, so we should make sure we add that step.

safern reviewed (commented):
This also needs to be done in runtime-official.yml

mangod9 committed:
PR Comments.

Removing compilerName parameter and renaming display names

mangod9 committed:
fix deletion typo

mangod9 commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,163 @@
+parameters:
+  archType: ''
+  buildConfig: ''
+  compilerName: 'clang'
```
removed compilername

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -1,7 +1,6 @@
-  compilerName: 'clang'
   condition: true
   container: ''
   crossrootfsDir: ''
```
crossrootfsDir is not needed either, right? I expect S.P.Corelib to only be OS specific, but not arch specific, so I expect we have only two legs, one for Windows and one for Unix.

trylek replied:
What about this bit in Array.CoreCLR.cs:

<pre>
#if TARGET_64BIT
using nuint = System.UInt64;
#else
using nuint = System.UInt32;
#endif
</pre>

safern replied:
Ah, you're right. I though it was not arch specific.

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -63,24 +57,7 @@ jobs:
-          value: '-clang6.0'
+      value: 'BuildLogs_CoreLib'
       # Building for x64 MUSL happens on Alpine Linux and we need to use the stable version available there
       - ${{ if and(eq(parameters.osGroup, 'Linux'), eq(parameters.osSubgroup, '_musl'), eq(parameters.archType, 'x64')) }}:
```
Not needed either.

safern replied:
Basically you can remove lines 61 through 77

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,124 @@
+    - script: $(Build.SourcesDirectory)$(dir)build$(scriptExt) -subsetCategory coreclr -subset corelib $(crossArg) -arch $(archType) -c $(buildConfig) $(officialBuildIdArg) -ci
+      displayName: Build System.Private.CoreLib only
+
+    # Save packages using the prepare-signed-artifacts format.
```
@jkoritzinsky is this needed?

jkoritzinsky replied:
I don't believe so. We should be able to remove this.

mangod9 replied:
this is removed, was actually causing an issue with official builds.

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,124 @@
+          installDotnet: true
+
+    # Build Private CoreLib
+    - script: $(Build.SourcesDirectory)$(dir)build$(scriptExt) -subsetCategory coreclr -subset corelib $(crossArg) -arch $(archType) -c $(buildConfig) $(officialBuildIdArg) -ci
```
Do we need the crossgened corelib as well? We usually run tests against nativecorelib. @jkoritzinsky ?

jkoritzinsky replied:
We'll crossgen CoreLib in the CoreCLR build job. The libraries build will get the IL CoreLib when compiling but the crossgened one when running tests.

safern replied:
I see. Does it make sense to build corelib twice? Doesn't it make more sense to just crossgen corelib here and use this job + the runtime job to run tests?

jkoritzinsky replied:
Crossgen isn't built here, so we can't run it.

If we want to just build it twice that's another option.

safern replied:
I think we should keep it as simple as possible as part of this PR so I'm fine with building the managed corelib here and then the native in Build CoreCLR legs and use those for testing as we're currently doing. Then we can improve the builds if needed.

mangod9 replied:
yeah agree, lets get the parallelization working and then we can optimize things further as needed.

mangod9 committed:
Adding step to upload output directory.

Also some cleanup

trylek commented:
Thanks Manish, you're making great progress here. As I see it based on our offline chat, as next step I would create another template, build-coreclr-job.yml, to basically contain (build-job.yml minus build-corelib-job.yml starting with corelib download from a previous corelib job) and change build-job.yml to just call these two templates. After that we can modify the template build-coreclr-and-libraries-job and adjust runtime and runtime-official as you've already started doing. As usual, we'll see what happens with the perf job that is somewhat special but I believe this should fix 90% of the existing pipelines.

trylek commented:
Hmm, you may be right that it's an overshoot to create yet another template and we can just "subtract" build-corelib-job from build-job. The only downside is that we need to audit all places in our pipelines calling build-job and update them accordingly. Keeping build-job as a wrapper for build-corelib-job followed by build-coreclr-job was supposed to maintain "backwards compatibility" during the transition.

mangod9 committed:
update libraries job to depend on corelib instead of product

mangod9 committed:
reverting libraries dependency change

Keeping the download of corelib artifacts to check if just corelib is sufficient to build libraires.

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -96,7 +96,7 @@ jobs:
         - ${{ if ne(parameters.liveRuntimeBuildConfig, '') }}:
           - _runtimeDownloadPath: '$(Build.SourcesDirectory)/artifacts/transport/${{ parameters.runtimeFlavor }}'
-          - _runtimeArtifactName: '$(runtimeFlavorName)Product_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveRuntimeBuildConfig }}'
+          - _runtimeArtifactName: 'CoreLib_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveRuntimeBuildConfig }}'
```
This will be braking for theor libraries/run-test-job.yml and also it is breaking the builds of libraries that build against mono.

I believe you will need to introduce an intermediate variable and define it in libraries/build-job, libraries/build-test-job and libraries/run-test-job.yml accordingly.

mangod9 replied:
fixed by checking for testScope

mangod9 committed:
Add OSX and musl_arm64 corelib builds + use full product for mono

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -96,7 +96,12 @@ jobs:
-          - _runtimeArtifactName: '$(runtimeFlavorName)Product_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveRuntimeBuildConfig }}'
+          # Use corelib dependencies for coreclr
+          - ${{ if ne(parameters.runtimeFlavor, 'mono') }}:
+            - _runtimeArtifactName: 'CoreLib_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveRuntimeBuildConfig }}'
```
I think this will break `libraries/run-test-job` for when runtimeFlavor 'coreclr'. I think we should add a variable that we set to: $(runtimeFlavorName)Product in `libraries/run-test-job` and then in `libraries/build-job` and `libraries/build-test-job` we set its value to `CoreLib_`.

Maybe the variable name can be `_coreclrArtifactPrefix`.

mangod9 replied:
now check for testScope to determine whether to use corelib or coreclr

mangod9 committed:
Another attempt to fix mono iOS case

mangod9 committed:
propagating runtimeFlavor to libraries base-job.yml

mangod9 committed:
updating DependsOn for libraries job to use corelib

mangod9 committed:
Merge remote-tracking branch 'upstream/master'

mangod9 committed:
Make LIbraries Test Build also depend on Product

mangod9 committed:
download product dependency when testScope is not null

mangod9 committed:
reverting a change which is not required

Hoping this fixes the test build.

mangod9 committed:
Adding runtime.depproj to libraries restore, and pretest.proj to tests.proj

pretest is required to generate the Microsoft.NETCore.App.deps files for CoreRun to work correctly.

mangod9 committed:
Fixing run_test_job to download product artifact

mangod9 committed:
run pretest.proj on test runs as well.

mangod9 committed:
Add libraries -restore to test run

since pretest needs installer.tasks.dll

mangod9 committed:
Updating pretest to use TargetArchitecture

A recent change caused this break :)

mangod9 committed:
Merge remote-tracking branch 'upstream/master'

mangod9 commented on src/libraries/restore/dirs.proj:
```diff
@@ -8,6 +8,7 @@
     <Project Include="netstandard/netstandard.depproj" />
     <Project Include="netfx/netfx.depproj" />
     <Project Include="tools/tools.depproj" />
+    <Project Include="runtime/runtime.depproj" />
```
this is breaking runtime-live-builds, since artifacts/bin/coreclr is required for this to work.

mangod9 replied:
fixed with this: https://github.com/dotnet/runtime/pull/34166/commits/e6d48826b13a0528323840b5bf951b40390baf5a

mangod9 committed:
Moving runtime restore from \restore\dirs.proj

Moved runtime restore + pretest.proj to teh tests.proj build.

mangod9 committed:
fix -allConfigurations build

mangod9 committed:
Adding corelib job to coreclr=and=Libraries.

mangod9 committed:
Add corelib to other pipelines.

mangod9 committed:
libraries test should depend on full product

This fix is similar to artifact download logic done in build-base.yml

mangod9 committed:
remove upload unsigned artifacts step from official build

mangod9 committed:
Merge remote-tracking branch 'upstream/master'

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,115 @@
+    condition: ${{ parameters.condition }}
+
+    name: ${{ format('coreclr_corelib_build_{0}{1}_{2}_{3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+    displayName: ${{ format('CoreLib Product Build {0}{1} {2} {3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
```
Nit, the displayName should be `CoreCLR CoreLib Product Build` as mono also has `Corelib` in it.

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,115 @@
+
+    # Build Private CoreLib
+    - script: $(Build.SourcesDirectory)$(dir)build$(scriptExt) -subsetCategory coreclr -subset corelib $(crossArg) -arch $(archType) -c $(buildConfig) $(officialBuildIdArg) -ci
+      displayName: Build System.Private.CoreLib only
```
Remove the trailing `only`.

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,115 @@
+    # Strip symbols only on the release build
+    - ${{ if eq(parameters.buildConfig, 'Release') }}:
+      - name: stripSymbolsArg
+        value: '-stripsymbols'
```
I believe this argument is no longer supported in the latest scripts. Might be worth rebasing before merging.

safern commented on eng/pipelines/coreclr/templates/build-corelib-job.yml:
```diff
@@ -0,0 +1,115 @@
+        tarCompression: $(tarCompression)
+        archiveExtension: $(archiveExtension)
+        artifactName: $(corelibProductArtifactName)
+        displayName: 'corelib build'
```
I would change the displayName to, `System.Private.CoreLib`.

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -95,7 +95,12 @@ jobs:
           - _runtimeDownloadPath: '$(Build.SourcesDirectory)/artifacts/transport/${{ parameters.runtimeFlavor }}'
-          - _runtimeArtifactName: '$(runtimeFlavorName)Product_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveRuntimeBuildConfig }}'
+          # Download corelib dependencies for coreclr libraries and not test
+          - ${{ if and(ne(parameters.runtimeFlavor, 'mono'), eq(parameters.testScope, '')) }}:
```
I rather change the condition here to check for `coreclr` as the `runtimeFlavor`.

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -49,7 +50,11 @@ jobs:
         - ${{ if ne(parameters.osGroup, 'WebAssembly') }}:
-          - ${{ format('{0}_product_build_{1}{2}_{3}_{4}', parameters.runtimeFlavor, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveRuntimeBuildConfig) }}
+          # Use corelib dependencies for coreclr
+          - ${{ if and(ne(parameters.runtimeFlavor, 'mono'), eq(parameters.testScope, '')) }}:
```
Ditto. Check for coreclr instead of not equal mono.

safern commented on eng/pipelines/libraries/build-test-job.yml:
```diff
@@ -38,6 +39,9 @@ jobs:
         - ${{ format('libraries_build_{0}_{1}{2}_{3}_{4}', parameters.framework, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+      # Libraries Test also depends on Product, now that the libraries build only depends on corelib
+      - ${{ if ne(parameters.liveRuntimeBuildConfig, '') }}:
+        - ${{ format('{0}_product_build_{1}{2}_{3}_{4}', parameters.runtimeFlavor, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveRuntimeBuildConfig) }}
```
I believe `allConfigurations` shouldn't depend on the whole coreclr build. It should only need corelib as the tests we run there don't use `testhost`.

safern replied:
Disregard, `allConfigurations` doesn't use this template at all. We can actually remove the `in(parameters.framework` conditions above and just leave it as:
```yml
- ${{ format('libraries_build_{0}{1}_{2}_{3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
# Libraries Test also depends on Product, now that the libraries build only depends on corelib
- ${{ if ne(parameters.liveRuntimeBuildConfig, '') }}:
   - ${{ format('{0}_product_build_{1}{2}_{3}_{4}', parameters.runtimeFlavor, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveRuntimeBuildConfig) }}
```

mangod9 replied:
ok will do as a subsequent PR.

mangod9 replied:
these are now included in this PR: https://github.com/dotnet/runtime/pull/34460

akoeplinger commented:
How hard would it be to extract Mono CoreLib too as part of this? It already builds completely independent from the Mono runtime so should just need .yml/.proj changes.

safern commented on eng/pipelines/libraries/outerloop.yml:
```diff
@@ -14,6 +14,29 @@ variables:
+  #
+  # CoreLib Build
+  #
+  - template: /eng/pipelines/common/platform-matrix.yml
```
To simplify this, we should be able to use `platform-matrix-multijob` on this yml. Not a blocker, I'll leave it up to you.

mangod9 replied:
will defer to next PR.

safern replied:
Sounds good.

safern commented on eng/pipelines/libraries/run-test-job.yml:
```diff
@@ -71,6 +72,9 @@ jobs:
             artifactName: $(librariesTestsArtifactName)
             artifactFileName: $(librariesTestsArtifactName)$(archiveExtension)
 
+        - script: $(_buildScript) -restore $(_buildArguments) $(_skipTestRestoreArg)
```
Please don't do a full restore. Why do you need to run restore? Is this to be able to run `pretest.proj`?

mangod9 replied:
i believe it needs installer.tasks to be restored at least.

safern replied:
Why is that needed? Running restore adds like 2 minutes to the build. So I want to understand why and see if we can add a workaround instead.

mangod9 replied:
because GenerateTestSharedFrameworkDepsFile task in pretest.proj is implemented there.

safern replied:
I see. We should be able to change  the command to do:

`$(_buildScript) -project src/libraries/pretest.proj ....`

That will ensure to use the root build script and will restore and build the installer tasks, but it will not do a full restore of the libraries subset.

mangod9 replied:
ok trying $(_buildScript) -project**s** src/libraries/pretest.proj now, also a fix to proj.test as suggested on chat.

safern commented on eng/pipelines/libraries/run-test-job.yml:
```diff
@@ -81,6 +85,16 @@ jobs:
                     /bl:$(Build.SourcesDirectory)/artifacts/log/$(_BuildConfig)/overrideCoreclrFromLiveDrop.binlog
             displayName: Override $(runtimeFlavorName) from live drop
 
+        - ${{ if ne(parameters.liveRuntimeBuildConfig, '') }}:
```
You don't need to duplicate this if condition, you can just use the one above, yml is indentation based.

mangod9 replied:
good catch.

safern commented on eng/pipelines/libraries/run-test-job.yml:
```diff
@@ -81,6 +85,16 @@ jobs:
+                    /p:RuntimeFlavor=${{ parameters.runtimeFlavor }}
+                    /p:TargetArchitecture=${{ parameters.archType }}
+                    /p:Configuration=${{ parameters.buildConfig }}
+                    /bl:$(Build.SourcesDirectory)/artifacts/log/$(_BuildConfig)/overrideCoreclrFromLiveDrop_pretest.binlog
```
the binarylog should be named different. This is not overridingCoreclr, this is just running pretest. Change to `pretest.binlog` only.

mangod9 replied:
fixed

safern commented on src/libraries/tests.proj:
```diff
@@ -43,6 +43,12 @@
   <Target Name="BuildAllProjects">
+    <!-- These projects are required to restore the runtime and setup testhost -->
+    <MSBuild Targets="Build"
+             Projects="$(MSBuildThisFileDirectory)restore\runtime\runtime.depproj;$(MSBuildThisFileDirectory)\pretest.proj"
```
I think this should only be done on CI. Locally we run `runtime.depproj` as part of the restore phase. 

I would add an `ItemGroup` to do this, something like:
```xml
<ItemGroup>
  <SetupProjects Include="$(MSBuildThisFileDirectory)\pretest.proj" />
  <SetupProjects Include="$(MSBuildThisFileDirectory)restore\runtime\runtime.depproj" />
</ItemGroup>

<MSBuild Targets="Build"
                 Projects="@(SetupProjects)
                Condition="'$(ContinuousIntegrationBuild)' == 'true'" 
                ... />
```

cc: @ViktorHofer

mangod9 replied:
done

safern commented:
@ViktorHofer might have insights as well as he has been working on re-shuffling pieces to be able to restore ref and src projects as part of the libraries build.

trylek reviewed (approved):
Looks great to me modulo Santi's remarks, thanks so much Manish for pulling this off!

mangod9 committed:
PR Comments

mangod9 commented on eng/pipelines/coreclr/templates/xplat-pipeline-job.yml:
```diff
@@ -75,6 +75,9 @@ jobs:
     - name: buildProductRootFolderPath
       value: '$(Build.SourcesDirectory)/artifacts/bin/coreclr/$(osGroup).$(archType).$(buildConfigUpper)'
 
+    - name: corelibProductArtifactName
```
for now the folder path corelib is same as coreclr

mangod9 committed:
typo fix :(

mangod9 commented:
> How hard would it be to extract Mono CoreLib too as part of this? It already builds completely independent from the Mono runtime so should just need .yml/.proj changes.

the main reasoning for this is to improve build times, it appears Mono builds are fairly fast -- would this help ?

safern commented:
> the main reasoning for this is to improve build times, it appears Mono builds are fairly fast -- would this help ?

Yeah, mono builds are fairly fast, I think the downsides of uploading/downloading assets and such, might not give us much gain, plus the only legs that build against mono are the iOS at the moment.

trylek commented:
I guess that, even with Mono build being fast, there would be some value in having both runtime builds behave the same including separate CoreLib build. Having said that, I believe that by all means this is worth a separate PR and perhaps even a separate design discussion.

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -96,7 +96,7 @@ jobs:
 
         - ${{ if ne(parameters.liveRuntimeBuildConfig, '') }}:
           - _runtimeDownloadPath: '$(Build.SourcesDirectory)/artifacts/transport/${{ parameters.runtimeFlavor }}'
-          - _runtimeArtifactName: '$(runtimeFlavorName)Product_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveRuntimeBuildConfig }}'
```
I think what @mangod9 has now makes sense, as for coreclr we need `corelib`  when not running/building tests and for mono we can just use the full Product build.

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -49,7 +50,11 @@ jobs:
+          - ${{ if or(ne(parameters.runtimeFlavor, 'coreclr'), ne(parameters.testScope, '')) }}:
+            - ${{ format('{0}_product_build_{1}{2}_{3}_{4}', parameters.runtimeFlavor, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveRuntimeBuildConfig) }}
         - ${{ if eq(parameters.osGroup, 'WebAssembly') }}:
           - ${{ format('{0}_product_build_{1}{2}_{3}_{4}', parameters.runtimeFlavor, 'linux', parameters.osSubgroup, 'x64', parameters.liveRuntimeBuildConfig) }}
```
WebAssembly is still depending on the full coreclr build. We should be able to also make it build against corelib by doing something similare as above.

mangod9 replied:
yup will do as a subsequent PR.

mangod9 replied:
these are now included in this PR: https://github.com/dotnet/runtime/pull/34460

safern commented on eng/pipelines/libraries/run-test-job.yml:
```diff
@@ -70,17 +71,26 @@ jobs:
             artifactName: $(librariesTestsArtifactName)
             artifactFileName: $(librariesTestsArtifactName)$(archiveExtension)
-
+          
```
Nit, adding extra indentation here.

mangod9 committed:
Fix run-test-job pretest build

safern commented on eng/pipelines/coreclr/ci.yml:
```diff
@@ -31,6 +31,52 @@ jobs:
+#
+# Debug Corelib builds
+#
+- template: /eng/pipelines/common/platform-matrix.yml
```
Now that I look at this job, we should just need to build corelib for checked as libraries build only use the checked runtime. 

https://github.com/dotnet/runtime/pull/34166/files#diff-b7c9a0014b41923da8701c3dbade7868R136

mangod9 replied:
yup fixed.

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -118,6 +118,46 @@ jobs:
+- template: /eng/pipelines/common/platform-matrix.yml
+  parameters:
+    jobTemplate: /eng/pipelines/coreclr/templates/build-corelib-job.yml
+    buildConfig: checked
```
We don't need the checked corelib job as libraries only build against a release corelib. It runs tests against a checked runtime, but we get corelib from the product build. We should be safe to remove this matrix.

mangod9 replied:
fixed.

safern reviewed (approved):
My last two comments are important to be addressed as part of this PR. After that, feel free to merge.

mangod9 committed:
removed some unused corelib configs.

safern reviewed (approved):
Thanks for working on this 😄

ViktorHofer commented on src/libraries/pretest.proj:
```diff
@@ -78,7 +78,8 @@
              BuildInParallel="true"
-             ContinueOnError="ErrorAndStop" />
+             ContinueOnError="ErrorAndStop" 
+             Condition="'$(SkipBuildProjects)' != 'true'"/>
```
What about instead directly invoking pretest.proj from CI and passing in the target that you care about?

safern replied:
Yeah that makes sense. We could just call change the current call to /t: GenerateTestSharedFrameworkDepsFile.

Maybe on a follow-up PR so that he can merge?

ViktorHofer replied:
Yeah I'm fine with that.

mangod9 replied:
ok will do on next PR.

ViktorHofer commented on src/libraries/tests.proj:
```diff
@@ -41,8 +41,19 @@
   <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />
   
+  <ItemGroup>
+    <SetupProjects Include="$(MSBuildThisFileDirectory)restore\runtime\runtime.depproj" />
```
This is CI only, why not directly invoking the runtime.depproj before tests.proj?

mangod9 replied:
one issue is that build parallelization might cause a race, pretest.proj has to complete before tests start to build, since pretest generates the dep file to ensure testhost is setup correctly.

ViktorHofer replied:
Not following. You can do two steps before the test build so the sequence would be:
1. runtime.proj
2. pretest.proj and
3. tests.proj

Noteworthy, if you would wait until Monday, we could just invoke the lib-pretest lib-tests subsets which would handle all that for you. I plan to merge this on Monday if ready: https://github.com/dotnet/runtime/pull/33553.

mangod9 replied:
I ideally want to merge this tomorrow, since this PR touches a bunch of infra pieces. I can commit to reworking this once your PR is merged.

ViktorHofer replied:
No problem at all, I can make the change afterwards.

mangod9 commented:
This is now merged, I will open another PR for couple of outstanding comments. Thanks everyone for your review and help on this.

ViktorHofer commented:
@mangod9 how did you merge this? Squash or merge? cc @jkotas

mangod9 commented:
regular merge, should it have been Squash?

jkotas commented:
Yep: https://github.com/dotnet/runtime/blob/master/docs/pr-guide.md#merging-pull-requests

Next time...

mangod9 commented:
oops sorry, should have checked on that.
